### PR TITLE
[LV] Create block in mask up-front if needed.

### PR DESCRIPTION
**Pull request by fhahn**
At the moment, block and edge masks are created on demand, which means that they are inserted at the point where they are demanded and then cached. It is possible that the mask for a block is looked up later at a point that's not dominated by the point where the mask has been inserted.

To avoid this, create masks up front on entry to the corresponding basic block and leave it to VPlan simplification to remove unneeded masks.

Note that we need to create masks for all blocks, if any of the blocks in the loop needs predication, as computing the mask of a block depends on the masks of its predecessor.

Needed for #76090.

### CODE REVIEW

**Review by ayalz — APPROVED** [review]
This looks good to me, thanks! Adding some minor nits, and thoughts:

One alternative may be to continue generating masks on-demand, but place them at top of block, e.g., by having createBlockInMask() bump its Builder (temporarily) to first non-phi position in block. Another alternative is to have (subsequent) users of a mask move it above them, if/when needed.
Note that even if one generates all masks eagerly, as first non-phi's in block, it's possible for other non-phi's to be inserted above them. In general, predication should be done as a one-shot VPlan-to-VPlan transform (according to roadmap), including all (needed) masks and their predicated users.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -8108,14 +8108,18 @@ void VPRecipeBuilder::createHeaderMask(VPlan &Plan) {
   BlockMaskCache[Header] = BlockMask;
 }
 
-VPValue *VPRecipeBuilder::createBlockInMask(BasicBlock *BB, VPlan &Plan) {
+VPValue *VPRecipeBuilder::getBlockInMask(BasicBlock *BB) {
   assert(OrigLoop->contains(BB) && "Block is not a part of a loop");
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Done, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -8108,14 +8108,18 @@ void VPRecipeBuilder::createHeaderMask(VPlan &Plan) {
...
 
   // Look for cached value.
   BlockMaskCacheTy::iterator BCEntryIt = BlockMaskCache.find(BB);
-  if (BCEntryIt != BlockMaskCache.end())
-    return BCEntryIt->second;
+  assert(BCEntryIt != BlockMaskCache.end());
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Added, thanks!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 3]
```diff
@@ -8758,13 +8762,20 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
...
   for (BasicBlock *BB : make_range(DFS.beginRPO(), DFS.endRPO())) {
     // Relevant instructions from basic block BB will be grouped into VPRecipe
     // ingredients and fill a new VPBasicBlock.
     if (VPBB != HeaderVPBB)
       VPBB->setName(BB->getName());
     Builder.setInsertPoint(VPBB);
 
+    if (VPBB != HeaderVPBB && NeedsMasks)
+      RecipeBuilder.createBlockInMask(BB, *Plan);
+
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 4]
```diff
@@ -138,8 +138,11 @@ class VPRecipeBuilder {
 
   /// A helper function that computes the predicate of the block BB, assuming
   /// that the header block of the loop is set to True or the loop mask when
-  /// tail folding. It returns the *entry* mask for the block BB.
-  VPValue *createBlockInMask(BasicBlock *BB, VPlan &Plan);
+  /// tail folding.
+  void createBlockInMask(BasicBlock *BB, VPlan &Plan);
+
+  // Returns the *entry* mask for the block BB.
+  VPValue *getBlockInMask(BasicBlock *BB);
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h]: Done, thanks. Also fixed the doc-comment

**Review by ayalz — COMMENTED** [review]
post-commit nits.

File: llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h [hunk 5]
```diff
@@ -138,8 +138,11 @@ class VPRecipeBuilder {
 
   /// A helper function that computes the predicate of the block BB, assuming
   /// that the header block of the loop is set to True or the loop mask when
-  /// tail folding. It returns the *entry* mask for the block BB.
-  VPValue *createBlockInMask(BasicBlock *BB, VPlan &Plan);
+  /// tail folding.
+  void createBlockInMask(BasicBlock *BB, VPlan &Plan);
+
+  /// Returns the *entry* mask for the block \p BB.
+  VPValue *getBlockInMask(BasicBlock *BB) const;
```
Comment (antoniofrighetto) [inline llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:145]: Shouldn't these two be `createMaskInBlock` and `getMaskInBlock` respectively?

**Pull request merged**